File: travaux_diriges/tp1/pi_mpi.py

#!/usr/bin/env python3
# pi_mpi.py
from mpi4py import MPI
import time, numpy as np

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

nb_samples_total = 40_000_000
nb_samples = nb_samples_total // size  # repartition uniforme des echantillons

# Debut de la mesure de temps (commence avant le calcul)
start = MPI.Wtime()

# Chaque processus initialise sa graine pour generer des nombres aleatoires differents
np.random.seed(int(time.time()) ^ rank)

# Generation des points aleatoires dans [-1, 1] pour x et y
x = np.random.uniform(-1, 1, nb_samples)
y = np.random.uniform(-1, 1, nb_samples)

# Compter le nombre de points tombant dans le cercle unite
local_count = np.count_nonzero(x*x + y*y <= 1.0)

# Reduire les resultats de tous les processus (somme)
total_count = comm.reduce(local_count, op=MPI.SUM, root=0)

# Fin de la mesure de temps
end = MPI.Wtime()

# Le processus maitre calcule l'approximation de pi et affiche le temps d'execution
if rank == 0:
    approx_pi = 4.0 * total_count / nb_samples_total
    print(f"Approximation de pi (mpi4py) = {approx_pi}")
    print(f"Temps d'execution = {end - start} secondes")

File: travaux_diriges/tp1/pi_mpi.c
        /* pi_mpi.c */
        #include <stdio.h>
        #include <stdlib.h>
        #include <time.h>
        #include <math.h>
        #include <mpi.h>
        
        int main(int argc, char *argv[]) {
            int rank, nbp;
            MPI_Init(&argc, &argv);
            MPI_Comm_rank(MPI_COMM_WORLD, &rank);
            MPI_Comm_size(MPI_COMM_WORLD, &nbp);
        
            unsigned long nbSamplesTotal = 10000000UL;
            /* Chaque processus traite une portion des echantillons */
            unsigned long nbSamples = nbSamplesTotal / nbp;
            unsigned long nbDarts_local = 0;
        
            /* Utiliser une graine differente pour chaque processus */
            unsigned int seed = time(NULL) ^ rank;
        
            double tdeb = MPI_Wtime();
            for (unsigned long i = 0; i < nbSamples; i++) {
                double x = (double)rand_r(&seed) / RAND_MAX * 2.0 - 1.0;
                double y = (double)rand_r(&seed) / RAND_MAX * 2.0 - 1.0;
                if (x*x + y*y <= 1.0)
                    nbDarts_local++;
            }
        
            unsigned long nbDarts_total = 0;
            MPI_Reduce(&nbDarts_local, &nbDarts_total, 1, MPI_UNSIGNED_LONG, MPI_SUM, 0, MPI_COMM_WORLD);
            double tfin = MPI_Wtime();
        
            if (rank == 0) {
                double ratio = (double)nbDarts_total / (nbSamples * nbp);
                double pi = 4.0 * ratio;
                printf("Approximation de pi (MPI) = %f\n", pi);
                printf("Temps d'execution (MPI) = %f secondes\n", tfin - tdeb);
            }
        
            MPI_Finalize();
            return 0;
        }  